### ISSUE TITLE
Dealing with partitions (and consumer groups)

### ISSUE THREAD

**Issue by akubisch** [dfdx/Kafka.jl#12]
Hi,

we're extending our work with Kafka and a microservice in Julia. Now we have for a topic "mytopic" three different partitions.

Very basically, I'm initializing my client using `kc = KafkaClient("kafka.my.localhost.de", 9092)` and try to get the latest offsets of my topics using `take!(latest_offset(kc, "mytopic", 0))`.

The following is weird: the topic is set up using three partitions. As far as I see it, it is not possible, yet, to let Kafka automatically choose partitions, is that true? When I try to go through partitions (i.e. additionally calling `take!(latest_offset(kc, "mytopic", 1))` and `take!(latest_offset(kc, "mytopic", 2))`), julia always hangs (or waits) unlimitedly for two of the partitions, while getting parts of the offsets in the third one. For other topics (all with three partitions), the "responsive" one might be different (sometimes 1, sometimes 2...). Is that a bug or am I not using the functionality correctly, yet?

In addition, is there a way to define a consumer group using Kafka.jl?

Thanks very much in advance!

**Comment by akubisch** [comment]
Additional information:
We are using three brokers. It seems that I do only get an answer of a partition, when it's on broker 1 and broker 1 is leader - under these constraints I can do anything on that partition, I'd like to. In any other case (i.e. different broker, leader != 1) every command hangs in an infinite loop. It also seems that fetching metadata only works in ~33% of all cases ;-) I guess it has to do with randomly assigned brokers!?

**Comment by dfdx** [comment]
Yes, it's most likely related to leader selection, but I need to check it out. I'll try to do it over the weekend.

**Comment by akubisch** [comment]
That would be great, thanks a lot!

**Comment by dfdx** [comment]
It seems like mistakes from early days of my Julia experience when Kafka.jl was created start catching me. `KafkaClient` creates a separate coroutine to handle responses from each broker's connection, but it turns out tasks are never actually switched and only one coroutine (connected only to one broker) actually reads data. 

Anyway I was going to rethink the whole async approach, so it's a good chance to do it. In `sync-client` branch I created a new type - `KClient` - which is similar to the current `KafkaClient` but is synchronous and doesn't use coroutines. The usage is as follows: 

```
kc = KClient(host, 9092)
offset = latest_offset(kc, "mytopic", 0)
messages = ...
produce(kc, "test", 0, messages)
offset_messages = fetch(kc, "test", 0, start_offset)
```
Note that there's no more channels and `take!`. I tested it on a topic with 3 partitions replicated to 3 brokers running locally on different ports, and it worked fine. If this works for you too and I have enough time, I plan to finish the client during the next week. 

Regarding consumer groups, no, they are currently not supported. Previously each distributed system I worked with (plain Java and Python consumers, Spark, Storm and others) stored consumer offsets differently, so I was delaying this task. However now Kafka provides consumer group API directly via brokers, so I'll take a look at what we can do with it. Do you have a particular use case for consumer groups?

**Comment by akubisch** [comment]
Hi,
thanks for the rapid great work - it runs perfectly, I get all info and messages on all partitions.

Consumer group use case: Basically I'd to make sure that: 1. if I replicate a microservice multiple times, each message is only consumed once (i.e. within the same consumer group) and that 2. this does not interfere with other microservices that might consume the same messages (different group).

**Comment by dfdx** [comment]
Just to make it explicit and ensure we are on the same page: this also implies that you need offset management, i.e. consumers should commit their offsets after reading to not start over after restart. In its turn, it implies that:

1. Each partition may be read by at most one consumer at the same time (otherwise each consumer will overwrite the other's offsets for this partition). If you have 3 partitions, at most 3 consumers (microservices) may read from the topic. 
2. Consumers should be rebalanced automatically when a new one is added. 

Kafka now supports these things in their [wire protocol](https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol), but it's quite complicated and involves lots of async logic. I'll create several issues for these features, but I'm not sure how much time it can take and when I'll be able to work on it.

**Comment by akubisch** [comment]
I agree completely, offset management would need to be done by Kafka (and this is as far as I understood also the current status of other APIs like in Java or Python). But I completely understand that this takes quite some time.

Regarding the sync-client branch I've observed some strange behaviour, but this could of course be caused by the quick implementation - when I try to start my service (i.e. run the script), I'm getting a lot of errors somehow caused by loading Kafka.jl (ending in `atexit hook threw an error: Base.MethodError(f=typeof(Core.Inference.convert)(), args=(Base.AssertionError, "invalid age range update"), world=0x0000000000000ac6)`). Probably you'd expect that, but just to make sure... When sourcing the script in Juno the error does not occur.

**Comment by dfdx** [comment]
No, it isn't expected and I can't see where it comes from: the message suggests that it's an issue with type inference inside Julia itself happening in `atexit` hook, but I don't think I use this hook in Kafka.jl. Can you make a minimal example of a script producing this error? Also, what version of Julia do you use?

**Comment by akubisch** [comment]
I'm using julia v0.6.2 (the JuliaPro bundled version). A minimal working example (as denoted above, error throws only via CLI, not in Juno):

```julia
using Kafka
using Weave
using DataFrames
using TimeSeries
using AWSS3

x=1
```

Note that the other libraries seem not to be important, just their number. Interestingly, when I load Kafka as the last library, the error does not occur:

```julia
using Weave
using DataFrames
using TimeSeries
using AWSS3
using Kafka

x=1
```

**Comment by dfdx** [comment]
Hm, I can't reproduce this error. I tried: 

 * Julia 0.6.0 and 0.6.2 (not JuliaPro, though)
 * from REPL and using `julia test.jl` 
 * both snippets - with Kafka.jl at the beginning and at the end 
 * all on my Ubuntu 16.04

Can you please try to run it on another version of Julia? It's quite possible you encountered a bug in the runtime itself.

**Comment by akubisch** [comment]
It seems like it really was in the runtime itself. I've had the bug in a docker container (and thus thought it should be perfectly reproducible ;-)), but not in the official julia one, but in a container which - during build - downloads the latest 0.6.2. Apparently within the past two days there have been some updates or bugfixes - I cannot reproduce the error anymore (except for my current JuliaPro installation). I will try to update (aka reinstall) JuliaPro and check afterwards, but I'm pretty sure it is not a Kafka.jl problem.

**Comment by dfdx** [comment]
Just FYI: last week I've been reading through Kafka docs and working on Kafka client redesign which should facilitate the change. Docs are sometimes unclear and I can't give any ETA, but I'll try to run it at least until some decision on design is taken.

**Comment by dfdx** [comment]
It turns out new consumer groups in Kafka require smart, very smart clients. Creating and maintaining such a client on a protocol level requires much more time than I have. Fortunately, from the time Kafka.jl was created general infrastructure evolved a lot, and so meet RDKafka.jl!

[RDKafka.jl](https://github.com/dfdx/RDKafka.jl) is a Julia wrapper around fully-featured and high-performance C/C++ client `librdkafka`. It's very basic at the moment, but already supports your use cases, e.g. producer:

```
p = KafkaProducer("localhost:9092")
partition = 0
produce(p, "mytopic", partition, "message key", "message payload")
```
and consumer using consumer groups and autocommiting offsets:

```
c = KafkaConsumer("localhost:9092", "my-consumer-group")
parlist = [("mytopic", 0), ("mytopic", 1), ("mytopic", 2)]
subscribe(c, parlist)
timeout_ms = 1000
for i=1:10
    msg = poll(String, String, c, timeout_ms)
    println(msg)
end
```
RDKafka includes build script for Linux. If you are on another OS, you'll need to build `librdkafka[.so, .dll, .dylib]` and update [search path](https://github.com/dfdx/RDKafka.jl/blob/master/src/init.jl#L4-L8) accordingly... or just provide me instructions for your OS so I could include them into repo.

**Comment by akubisch** [comment]
This is great news, thank you! I think that's a very sensible approach. I'll have a look at RDKafka and try it out.
